mlir/lib/Quantum/Utils/QuantumSplitting.cpp

```diff
@@ -89,17 +91,95 @@ void AugmentedCircuitGenerator::generate(Region &region, OpBuilder &builder)
-            ValueRange diffParams = gate.getDiffParams();
-            if (!diffParams.empty()) {
-                for (Value param : diffParams) {
-                    builder.create<ListPushOp>(gate.getLoc(), oldToCloned.lookupOrDefault(param),
-                                               cache.paramVector);
+        else if (auto gate = dyn_cast<quantum::ParametrizedGate>(op)) {
+            ValueRange params = gate.getAllParams();
+
+            for (Value param : params) {
+                Location loc = gate.getLoc();
+                Value clonedParam = oldToCloned.lookupOrDefault(param);
+                Type paramType = clonedParam.getType();
+
+                if (paramType.isF64()) {
+                    builder.create<ListPushOp>(loc, clonedParam, cache.paramVector);
+                    continue;
+                }
+
+                // Sanitizing inputs.
+                // Technically we know for a fact that none of this will ever issue an error.
+                // This is because QubitUnitary is guaranteed to have a tensor<NxNxcomplex<f64>>
+                // But this code in the future may be extended to support other types.
+                // Hence the sanitization.
+                if (!isa<RankedTensorType>(paramType)) {
+                    gate.emitOpError() << "Unexpected type.";
+                }
+
+                auto aTensor = paramType.cast<RankedTensorType>();
+                ArrayRef<int64_t> shape = aTensor.getShape();
+
+                if (2 != shape.size()) {
+                    gate.emitOpError() << "Unexpected tensor shape in QubitUnitaryOp";
+                }
+                if (shape[0] != shape[1]) {
+                    gate.emitOpError() << "QubitUnitaryOp is not square matrix.";
+                }
+                if (shape[0] < 2) {
+                    gate.emitOpError() << "QubitUnitaryOp has invalid tensor shape.";
+                }
+                int64_t n = shape[0];
+                bool isPowerOfTwo = ((n & (n - 1)) == 0);
+                if (!isPowerOfTwo) {
+                    gate.emitOpError() << "QubitUnitaryOp is not of a valid size.";
+                }
+                Type elementType = aTensor.getElementType();
+                if (!isa<ComplexType>(elementType)) {
+                    gate.emitOpError() << "QubitUnitaryOp does not hold complex types.";
+                }
+                Type f64 = elementType.cast<ComplexType>().getElementType();
+                if (!f64.isF64()) {
+                    gate.emitOpError() << "QubitUnitaryOp does not hold complex F64 types.";
+                }
+
+                // Constants
+                Value c0 = builder.create<index::ConstantOp>(loc, 0);
+                Value c1 = builder.create<index::ConstantOp>(loc, 1);
+                // Matrix of size NxN
+                // we can use either shape[0] or shape[1] because matrix is square
+                Value N = builder.create<index::ConstantOp>(loc, shape[0]);
```

Comment:
Can we add support for a dynamic shape here? If the dimsize is static, create the constant, otherwise get it from the matrix value (`tensor.dim`)?